### PR TITLE
Refactor the Credential Re-Up Code.

### PR DESCRIPTION
**Pull request by diranged**
Rather than waiting an explicit number of seconds, check every 15
seconds if the creds have expired or not (without an API call). When the
creds are 10m away from expiring, go and actively get new ones.

This makes it easier for us to handle network connection blips (closed
laptop, etc) and reduces the complexity for the user.

Also in here is a slight change to the way the Session object works --
treating it more like a real object (with properties, etc).

### CODE REVIEW

File: nd_okta_auth/aws.py [hunk 1]
```diff
@@ -114,8 +116,23 @@ def __init__(self,
...
+        self.aws_session_token = None
+        self.expiration = None
+
+    @property
+    def is_valid(self):
+        '''Returns True if the Session is still valid.'''
+        now = time.mktime(datetime.datetime.utcnow().timetuple())
+        expir = time.mktime(self.expiration.timetuple())
+
+        return (now - 600) < expir
```
Comment (stlava) [inline nd_okta_auth/aws.py]: Magic number. Also why is the offset needed?

File: nd_okta_auth/main.py [hunk 2]
```diff
@@ -152,29 +152,35 @@ def main(argv):
...
-            log.error(e)
-        except aws.BaseException as e:
-            log.error('Exception in SAML Assertion')
-            log.error(e)
-        finally:
+            assertion = okta_client.get_assertion(appid=config.appid,
+                                                  apptype='amazon_aws')
+            session = aws.Session(assertion, profile=config.name)
+            session.assume_role()
+        except requests.exceptions.ConnectionError as e:
```
Comment (stlava) [inline nd_okta_auth/main.py]: Which line does this `except handle`? Can you maybe break it up so you only do 1 action per try block? Are there other non-fatal exceptions we can gracefully handle with a retry?

File: nd_okta_auth/main.py [hunk 3]
```diff
@@ -152,29 +152,35 @@ def main(argv):
...
+            log.warning('Connection error... will retry')
             time.sleep(5)
+            continue
+
+        # If we're not running in re-up mode, once we have the assertion
+        # and creds, go ahead and quit.
+        if not config.reup:
+            break
+
+        time.sleep(5)
```
Comment (stlava) [inline nd_okta_auth/main.py]: I think this sleep is not needed.

File: nd_okta_auth/main.py [hunk 4]
```diff
@@ -152,29 +152,35 @@ def main(argv):
...
     # object to get a fresh SAMLResponse repeatedly and refresh our AWS
     # Credentials.
+    session = None
     while True:
+        # If an AWS Session object has been created already, lets check if its
+        # still valid. If it is, sleep a bit and skip to the next execution of
+        # the loop.
+        if session and session.is_valid:
+            log.debug('Credentials are still valid, sleeping')
+            time.sleep(15)
```
Comment (stlava) [inline nd_okta_auth/main.py]: This can just sleep until shortly before expiration. No need for a loop.

Comment (diranged) [inline nd_okta_auth/main.py]: Fair -- that requires this to have some more logic in it to understand when its going to expire.

File: nd_okta_auth/main.py (hunk 2 above)
Comment (diranged) [inline nd_okta_auth/main.py]: It actually handles the top and bottom line inside of the block..

**Comment by diranged** [comment]
@stlava found a bug in testing.. fixed the bug. will write tests later to catch this bug in the future. :P

**Comment by diranged** [comment]
@Stlava quick final review?

File: nd_okta_auth/aws.py [hunk 5]
```diff
@@ -114,8 +116,41 @@ def __init__(self,
...
+            buffer: Number of seconds before the actual expiration before we
+            start returning false.
+
+        Returns:
+            Bool
+        '''
+        # Consider the tokens expired when they have 10m left
+        buffer = 600
+
+        now = time.mktime(datetime.datetime.utcnow().timetuple())
```
Comment (stlava) [inline nd_okta_auth/aws.py]: nit to make this a little cleaner
```
buffer = datetime.timedelta(seconds=600)
now =  datetime.datetime.utcnow()
expir = datetime.datetime.strptime(str(self.expiration), '%Y-%m-%d %H:%M:%S+00:00')

return (now + buffer) < expir
```

File: nd_okta_auth/aws.py [hunk 6]
```diff
@@ -114,8 +116,41 @@ def __init__(self,
 
         self.profile = profile
         self.region = region
+
         self.assertion = models.SamlAssertion(assertion)
-        self.credentials = Credentials(cred_file)
+        self.writer = Credentials(cred_file)
```
Comment (stlava) [inline nd_okta_auth/aws.py]: Does `Credentials()` throw any errors that need catching here?

**Pull request merged**